server/backends/blobstore/gcs/gcs.go

```diff
@@ -200,35 +200,38 @@ func (g *GCSBlobStore) BlobExists(ctx context.Context, blobName string) (bool, e
 	}
 }
 
+// haveBeenSet returns a boolean indicating if the set of preconditions is empty
+// or not. If any have been set, true is returned, otherwise the preconditions
+// may be assumed to be empty.
+func haveBeenSet(conds storage.Conditions) bool {
+	switch {
+	case conds.GenerationMatch != 0:
+		return true
+	case conds.GenerationNotMatch != 0:
+		return true
+	case conds.DoesNotExist:
+		return true
+	case conds.MetagenerationMatch != 0:
+		return true
+	case conds.MetagenerationNotMatch != 0:
+		return true
+	default:
+		return false
+	}
+}
+
 func (g *GCSBlobStore) ConditionalWriter(ctx context.Context, blobName string, conds storage.Conditions) (interfaces.CommittedWriteCloser, error) {
 	ctx, cancel := context.WithCancel(ctx)
-	bw := g.bucketHandle.Object(blobName).If(conds).NewWriter(ctx)
-
-	var zw io.WriteCloser
-	if g.compress {
-		zw = util.NewCompressWriter(bw)
+	var bw *storage.Writer
+	if haveBeenSet(conds) {
+		bw = g.bucketHandle.Object(blobName).If(conds).NewWriter(ctx)
 	} else {
-		zw = bw
-	}
-	cwc := ioutil.NewCustomCommitWriteCloser(zw)
-	cwc.CommitFn = func(int64) error {
-		if compresserCloseErr := zw.Close(); compresserCloseErr != nil {
-			cancel() // Don't try to finish the commit op if Close() failed.
-			// Canceling the context closes the Writer, so don't call bw.Close().
-			return compresserCloseErr
-		}
-		return bw.Close()
-	}
-	cwc.CloseFn = func() error {
-		cancel()
-		return nil
+		bw = g.bucketHandle.Object(blobName).NewWriter(ctx)
 	}
-	return cwc, nil
-}
 
-func (g *GCSBlobStore) Writer(ctx context.Context, blobName string) (interfaces.CommittedWriteCloser, error) {
-	ctx, cancel := context.WithCancel(ctx)
-	bw := g.bucketHandle.Object(blobName).NewWriter(ctx)
+	// See https://pkg.go.dev/cloud.google.com/go/storage#Writer
+	// Always disable buffering in the client for these writes.
+	bw.ChunkSize = 0
```

Comment:
from the gcs docs:
```
	// If ChunkSize is set to zero, chunking will be disabled and the object will
	// be uploaded in a single request without the use of a buffer. This will
	// further reduce memory used during uploads, but will also prevent the writer
	// from retrying in case of a transient error from the server or resuming an
	// upload that fails midway through, since the buffer is required in order to
	// retry the failed request.
```
is there some higher-level retry logic i'm missing to compensate for this? (it didn't look like it to me, but i'm also not sure we should care?)

Reply:
I didn't add anything, but given that we're already disabling this for chunks under 16M (most chunks) it seems fairly safe. LMK if you disagree tho, not attached to this part if you feel strongly.